src/metatrain/experimental/nanopet/trainer.py

```diff
@@ -0,0 +1,481 @@
+        old_lr = optimizer.param_groups[0]["lr"]
+        logger.info(f"Initial learning rate: {old_lr}")
+
+        start_epoch = 0 if self.epoch is None else self.epoch + 1
+
+        @torch.jit.script
+        def systems_and_targets_to_device(
+            systems: List[System], targets: Dict[str, TensorMap], device: torch.device
+        ) -> Tuple[List[System], Dict[str, TensorMap]]:
+            return (
+                [system.to(device=device) for system in systems],
+                {key: value.to(device=device) for key, value in targets.items()},
+            )
+
+        @torch.jit.script
+        def systems_and_targets_to_dtype(
+            systems: List[System], targets: Dict[str, TensorMap], dtype: torch.dtype
+        ) -> Tuple[List[System], Dict[str, TensorMap]]:
+            return (
+                [system.to(dtype=dtype) for system in systems],
+                {key: value.to(dtype=dtype) for key, value in targets.items()},
+            )
+
+        # Train the model:
+        if self.best_loss is None:
+            self.best_loss = float("inf")
+        logger.info("Starting training")
+        for epoch in range(start_epoch, start_epoch + self.hypers["num_epochs"]):
+            if is_distributed:
+                sampler.set_epoch(epoch)
+
+            train_rmse_calculator = RMSEAccumulator()
+            val_rmse_calculator = RMSEAccumulator()
+            if self.hypers["log_mae"]:
+                train_mae_calculator = MAEAccumulator()
+                val_mae_calculator = MAEAccumulator()
+
+            train_loss = 0.0
+            for batch in train_dataloader:
+                optimizer.zero_grad()
+
+                systems, targets = batch
+                systems, targets = apply_random_augmentations(systems, targets)
+                systems, targets = systems_and_targets_to_device(
+                    systems, targets, device
+                )
+                for additive_model in (
+                    model.module if is_distributed else model
+                ).additive_models:
+                    targets = remove_additive(
+                        systems, targets, additive_model, train_targets
+                    )
+                systems, targets = systems_and_targets_to_dtype(systems, targets, dtype)
+                predictions = evaluate_model(
+                    model,
+                    systems,
+                    {key: train_targets[key] for key in targets.keys()},
+                    is_training=True,
+                )
+
+                # average by the number of atoms
+                predictions = average_by_num_atoms(
+                    predictions, systems, per_structure_targets
+                )
+                targets = average_by_num_atoms(targets, systems, per_structure_targets)
+
+                train_loss_batch = loss_fn(predictions, targets)
+                train_loss_batch.backward()
+                torch.nn.utils.clip_grad_norm_(model.parameters(), 1.0)
```

Comment:
Do we always want to have gradient clipping? I had an impression that it sometimes work in very unpredictable way

Reply:
That's a good point. I can try to experiment with it and make it an option if I see some instability

Reply:
Since this parameter is supposed to be near-optimal most of the time, I would keep it as hardcoded. Since this is supposed to be a "clean" version, I wouldn't want to have too many parameters that are not needed